### PR TITLE
Optimization of buffer size for reading from SmartFs files

### PR DESCRIPTION
**Pull request by amogh-samsung**
- Using just any buffer size for reading from SmartFs files causes
  repetitive reading of the same sector and degrades performance.
- We suggest the use of an optimal buffer size for reading from
  SmartFs files when the content is to be read back-to-back in a loop.

### CODE REVIEW

File: os/fs/driver/mtd/smart.c [hunk 1]
```diff
@@ -5854,6 +5854,17 @@ static int smart_journal_recovery(FAR struct smart_struct_s *dev, journal_log_t
 
 #endif
 
+/****************************************************************************
+ * Name: get_smart_mtd_header_size
+ *
+ * Description: Return the size of the smart_sect_header_structure
+ *
+ ****************************************************************************/
+int get_smart_mtd_header_size()
```
Comment (jeongchanKim) [inline os/fs/driver/mtd/smart.c:5863]: We usually name the function as MODULENAME_VERB_OBJECTIVES.
So I think `smart_get_smart_mtd_header_size` or `smart_get_mtd_header_size` is better.

File: os/fs/smartfs/smartfs.h [hunk 2]
```diff
@@ -434,4 +434,16 @@ struct fs_dirent_s;
...
+ * Public function prototypes
+ ****************************************************************************/
+
+/****************************************************************************
+ * Description:
+ *   This function returns the optimal read buffer size for back-to-back
+ *   reading from SmartFs i.e. number of actual data bytes in the sector.
+ *
+ ****************************************************************************/
+int get_smartfs_optimal_read_size(void);
```
Comment (jeongchanKim) [inline os/fs/smartfs/smartfs.h:447]: Same as above.

File: os/fs/smartfs/smartfs_smart.c [hunk 3]
```diff
@@ -1738,3 +1738,17 @@ static int smartfs_stat(struct inode *mountpt, const char *relpath, struct stat
...
+ *
+ * Description: Return the optimal buffer size for reading from SmartFs to
+ *   avoid repetitive reading of sectors. This size is the number of actual
+ *   data bytes in the sector
+ *
+ * **************************************************************************/
+
+int get_smartfs_optimal_read_size()
+{
+	return (CONFIG_MTD_SMART_SECTOR_SIZE - sizeof(struct smartfs_chain_header_s) - get_smart_mtd_header_size());
```
Comment (Taejun-Kwon) [inline os/fs/smartfs/smartfs_smart.c:1753]: block driver layer. should be hided to fs layer, this can' be accepted. you should use ioctl

**Comment by Taejun-Kwon** [comment]
We already have a method to get real buffer size from lower level of file system to above layer, framework or app.
Please see init_test_file() in smartfs_powercut_main.c

File: os/fs/driver/mtd/smart.c (hunk 1 above)
Comment (davidfather) [reply os/fs/driver/mtd/smart.c:5863]: @amogh-samsung I also agree with this suggestion.

File: os/fs/smartfs/smartfs_smart.c (hunk 3 above)
Comment (davidfather) [reply os/fs/smartfs/smartfs_smart.c:1753]: @amogh-samsung With the loadable configuration, this API can't be called from applications unless a new ioctl interface is introduced.

File: apps/README.md [hunk 4]
```diff
@@ -65,3 +66,13 @@ CONFIGURED_APPS += examples/hello
 endif
 ```
 
+## Optimized Buffering for SmartFs READ
+
+Using just any buffer size for reading the contents of a SmartFs file, back-to-back,
```
Comment (davidfather) [inline apps/README.md:71]: @amogh-samsung Could you please make another commit within this PR, which can test this new API. This test code might be seen as an example.

File: os/fs/smartfs/smartfs_smart.c (hunk 3 above)
Comment (amogh-samsung) [reply os/fs/smartfs/smartfs_smart.c]: Updated

File: apps/README.md (hunk 4 above)
Comment (amogh-samsung) [reply apps/README.md]: Added

File: apps/examples/smartfs/read_optimization/Makefile [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+# smartfs powercut Test built-in application info
+
+APPNAME = smartfs_read_opt
+FUNCNAME = read_optimization_main
+PRIORITY = SCHED_PRIORITY_DEFAULT
+STACKSIZE = 4096
+THREADEXEC = TASH_EXECMD_ASYNC
+
+# smartfs powercut Test Example
```
Comment (dineshdas-samsung) [inline apps/examples/smartfs/read_optimization/Makefile:65]: Can we Please change the comment here ?

File: apps/examples/smartfs/read_optimization/Makefile [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+############################################################################
+
+-include $(TOPDIR)/.config
+-include $(TOPDIR)/Make.defs
+include $(APPDIR)/Make.defs
+
+# smartfs powercut Test built-in application info
```
Comment (dineshdas-samsung) [inline apps/examples/smartfs/read_optimization/Makefile:57]: Can we Please change the comment here ?

**Comment by amogh-samsung** [comment]
**>>>>>>>>>>>>>>>>>>>>>SECTOR SIZE 1024<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<**


> TASH>>smartfs_read_opt 1
> TASH>>
> **************************************************************************************
> Time ten to read 8Kb file with 512 byte buffer, 100 times = 643290
> **************************************************************************************
> 
> **************************************************************************************
> Time taken to read 8Kb file with 1024 byte buffe 100 times = 509040
> 
> **************************************************************************************
> 
> **************************************************************************************
> Time taken to read 8Kb file with 13 byte buffer, 100 times = 410490
> 
> **************************************************************************************
> Test Example Complete
> 
> TASH>>
> TASH>>
> TASH>>smartfs_read_opt 1 20
> TASH>>
> **************************************************************************************
> Time taken to read 20Kb file with 512 byte buffer, 100imes = 1487280
> 
> **************************************************************************************
> 
> **************************************************************************************
> Time taken to read 20Kb file with 1024 byte buffer, 100 times = 1150740
>                                                                                              
> **************************************************************************************         
>                                                                                              
> **************************************************************************************       
> Time taken to read 20Kb file with 10 byte buffer, 100 times = 884490
> 
> **************************************************************************************
> Test Example Complete
> 
> TASH>>
> TASH>>
> TASH>>smartfs_read_opt 1 10
> TASH>>
> **************************************************************************************
> Time taken to rea10Kb file with 512 byte buffer, 100 times = 786810
> 
> **************************************************************************************
> 
> **************************************************************************************
> Time taken to read 10Kb file with 1024 byte buffer, 100 times = 617610
> 
> **************************************************************************************
> 
> **************************************************************************************
> Time taken to read 10Kb file with 13 byte buffer, 100 times = 491400
> 
> **************************************************************************************
> Test Example Complete



 **>>>>>>>>>>>>>>>>>>>>>SECTOR SIZE 512<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<**

> TASH>>smartfs_read_opt 1
> TASH>>
> **************************************************************************************
> Time taken to read 8Kb file with 512 byte buffer, 100 tes = 529050
> 
> **************************************************************************************
> 
> **************************************************************************************
> Time taken to read 8Kb file with 1024 be buffer, 100 times = 451140
> 
> **************************************************************************************
> 
> **************************************************************************************
> Time taken to read 8Kb file with 5 byte buffer, 100 times = 405360
> 
> **************************************************************************************
> Test Example Complete
> 
> TASH>>
> TASH>>
> TASH>>smartfs_read_opt 1 20
> TASH>>
> **************************************************************************************
> Time taken to read 20Kb file with 2 byte buffer, 100 times = 1238700
> 
> **************************************************************************************
> 
> **************************************************************************************
> Time taken to read 20Kb file with 1024 byte buffer, 100 times 1050960
> 
> **************************************************************************************       
>                                                                                              
> **************************************************************************************
> Time taken to read0Kb file with 501 byte buffer, 100 times = 925230                          
> 
> **************************************************************************************
> Test Example Complete
> 
> TASH>>
> TASH>>
> TASH>>smartfs_read_opt 1 10
> TASH>>
> **************************************************************************************
> Time taken to read 10Kb file wh 512 byte buffer, 100 times = 644790
> 
> **************************************************************************************
> 
> **************************************************************************************
> Time taken to read 10Kb file with 24 byte buffer, 100 times = 550800
> 
> **************************************************************************************
> 
> **************************************************************************************
> Time tak to read 10Kb file with 501 byte buffer, 100 times = 491970
> **************************************************************************************
> Test Example Complete

File: apps/examples/smartfs/read_optimization/read_optimization_main.c [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+ * Description: Creates the test file with test data that we will use
+ *              to conduct the test.
+ *
+ ****************************************************************************/
+
+static int create_test_file(int f_size)
+{
+	int fd;
+	int ret;
+	memset(g_buf, 1024, '1');
```
Comment (dineshdas-samsung) [inline apps/examples/smartfs/read_optimization/read_optimization_main.c:126]: Did we mean memset(g_buf, '1', 1024); ?

File: apps/examples/smartfs/read_optimization/read_optimization_main.c [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+	time_read = time_2.timeleft - time_1.timeleft;
+	printf("\n****************************************************************************************\n");
+	printf("Time taken to read %dKb file with %d byte buffer, 100 times = %lu\n", size, buf_size, time_read);
+	printf("\n****************************************************************************************\n");
+
+	if (ioctl(frt_fd, TCIOC_STOP, 0) < 0) {
+		fprintf(stderr, "ERROR: Failed to stop the Free Run Timer: %d\n", errno);
+	}
+
+	unlink(SAMPLE_FILE);
```
Comment (dineshdas-samsung) [inline apps/examples/smartfs/read_optimization/read_optimization_main.c:269]: Can we add a check here also please whether or not unlink fails ?

File: apps/examples/smartfs/read_optimization/read_optimization_main.c (hunk 7 above)
Comment (amogh-samsung) [reply apps/examples/smartfs/read_optimization/read_optimization_main.c]: Done

File: apps/examples/smartfs/read_optimization/Makefile (hunk 5 above)
Comment (amogh-samsung) [reply apps/examples/smartfs/read_optimization/Makefile]: Done

File: apps/examples/smartfs/read_optimization/Makefile (hunk 6 above)
Comment (amogh-samsung) [reply apps/examples/smartfs/read_optimization/Makefile]: Done

File: apps/examples/smartfs/read_optimization/read_optimization_main.c (hunk 8 above)
Comment (amogh-samsung) [reply apps/examples/smartfs/read_optimization/read_optimization_main.c:269]: We can work without it since it's only a sample application here.

**Review by Taejun-Kwon — COMMENTED** [review]
Instead of read_opimizetion.main.c, I think smartfs_performance_test.c is better.
If we use st_blksize as a buffer size, performance of write also will become more good.
and it can guarantee integrity of written data too.
That's why I used 2 type of buffer size in powercut test....

**Comment by davidfather** [comment]
> Instead of read_opimizetion.main.c, I think smartfs_performance_test.c is better. If we use st_blksize as a buffer size, performance of write also will become more good. and it can guarantee integrity of written data too. That's why I used 2 type of buffer size in powercut test....

@amogh-samsung What TJ suggested might be a good solution to improve write throughput. Please check this.

**Comment by davidfather** [comment]
> **>>>>>>>>>>>>>>>>>>>>>SECTOR SIZE 1024<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<**
> 
> > TASH>>smartfs_read_opt 1
> > TASH>>
> > 
> > Time ten to read 8Kb file with 512 byte buffer, 100 times = 643290
> > 
> > 
> > Time taken to read 8Kb file with 1024 byte buffe 100 times = 509040
> > 
> > 
> > Time taken to read 8Kb file with 13 byte buffer, 100 times = 410490
> > 
> > Test Example Complete
> > TASH>>
> > TASH>>
> > TASH>>smartfs_read_opt 1 20
> > TASH>>
> > 
> > Time taken to read 20Kb file with 512 byte buffer, 100imes = 1487280
> > 
> > 
> > Time taken to read 20Kb file with 1024 byte buffer, 100 times = 1150740
> > 
> > 
> > Time taken to read 20Kb file with 10 byte buffer, 100 times = 884490
> > 
> > Test Example Complete
> > TASH>>
> > TASH>>
> > TASH>>smartfs_read_opt 1 10
> > TASH>>
> > 
> > Time taken to rea10Kb file with 512 byte buffer, 100 times = 786810
> > 
> > 
> > Time taken to read 10Kb file with 1024 byte buffer, 100 times = 617610
> > 
> > 
> > Time taken to read 10Kb file with 13 byte buffer, 100 times = 491400
> > 
> > Test Example Complete
> 
> **>>>>>>>>>>>>>>>>>>>>>SECTOR SIZE 512<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<**
> 
> > TASH>>smartfs_read_opt 1
> > TASH>>
> > 
> > Time taken to read 8Kb file with 512 byte buffer, 100 tes = 529050
> > 
> > 
> > Time taken to read 8Kb file with 1024 be buffer, 100 times = 451140
> > 
> > 
> > Time taken to read 8Kb file with 5 byte buffer, 100 times = 405360
> > 
> > Test Example Complete
> > TASH>>
> > TASH>>
> > TASH>>smartfs_read_opt 1 20
> > TASH>>
> > 
> > Time taken to read 20Kb file with 2 byte buffer, 100 times = 1238700
> > 
> > 
> > Time taken to read 20Kb file with 1024 byte buffer, 100 times 1050960
> > 
> > 
> > Time taken to read0Kb file with 501 byte buffer, 100 times = 925230
> > 
> > Test Example Complete
> > TASH>>
> > TASH>>
> > TASH>>smartfs_read_opt 1 10
> > TASH>>
> > 
> > Time taken to read 10Kb file wh 512 byte buffer, 100 times = 644790
> > 
> > 
> > Time taken to read 10Kb file with 24 byte buffer, 100 times = 550800
> > 
> > 
> > Time tak to read 10Kb file with 501 byte buffer, 100 times = 491970
> > 
> > Test Example Complete

@amogh-samsung Please, check the buffer size of

File: apps/examples/smartfs/read_optimization/read_optimization_main.c [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+	read_test(512);
+
+	if (ioctl(frt_fd, TCIOC_GETSTATUS, (unsigned long)(uintptr_t)&time_2) < 0) {
+		fprintf(stderr, "ERROR: Failed to get Free Run Timer status: %d\n", errno);
+		return -1;
+	}
+
+	time_read = time_2.timeleft - time_1.timeleft;
+	printf("\n****************************************************************************************\n");
+	printf("Time taken to read %dKb file with 512 byte buffer, 100 times = %lu\n", size, time_read);
```
Comment (davidfather) [inline apps/examples/smartfs/read_optimization/read_optimization_main.c:221]: For better readability, could you include the throughput of each test, instead of the total time taken to read a file?

File: apps/README.md [hunk 10]
```diff
@@ -65,3 +66,19 @@ CONFIGURED_APPS += examples/hello
 endif
 ```
 
+## Suggestion for optimizing SmartFs Read
+
+USing just any buffer size for reading the contents of a SmartFs file, back-to-back(looped),
```
Comment (abhishek-samsung) [inline apps/README.md:71]: Please change USing --> Using

Comment (amogh-samsung) [reply apps/README.md:71]: Done

File: apps/examples/smartfs/read_optimization/read_optimization_main.c (hunk 9 above)
Comment (amogh-samsung) [reply apps/examples/smartfs/read_optimization/read_optimization_main.c:221]: Updated

**Comment by amogh-samsung** [comment]
Updated the example to display thoughput for the read operation instead of time. Output looks like the following now: -

> TASH>>smartfs_read_opt 1
> TASH>>Timer path = /dev/timer1
> 
> ***************************************************************************************
> Throughput while reading 8Kb file with 512 byte buffer = 16.72Kbps
> 
> **************************************************************************************
> 
> ****************************************************************************************
> Throughput while reading 8Kb file with 1024 byte buffer = 19.69Kbps
> 
> **************************************************************************************
> 
> **************************************************************************************
> Throughput while reading 8Kb file with 501 byte buffer = 21.94Kbps
> 
> ************************************************************************************
> Test Example Complete
